src/crimson/os/seastore/onode_manager/staged-fltree/node_extent_accessor.h

```diff
@@ -174,32 +186,47 @@ class DeltaRecorderT final: public DeltaRecorder {
   }
 
  private:
-  static void encode_value(const value_input_t& value, ceph::bufferlist& encoded) {
+  ValueDeltaRecorder* get_value_replayer(value_magic_t magic) {
+    // Replay procedure is independent of Btree and happens at lower level in
+    // seastore. There is no ValueBuilder so the recoder needs to build the
+    // ValueDeltaRecorder by itself.
+    if (!value_replayer) {
+      value_replayer = build_value_recorder_by_type(encoded, magic);
+      if (!value_replayer) {
+        logger().error("OTree::Extent::Replay: got unexpected value magic={}",
+                       magic);
+        ceph_abort();
+      }
+    }
+    if (value_replayer->get_header_magic() != magic) {
```

Comment:
i think `build_value_recorder_by_type()` verifies this already.

Reply:
No, this is to validate whether the value header magic matches the magic of the already loaded `value_replayer`. `build_value_recorder_by_type()` is only called at the first time when `value_replay` is not available in this node.

Reply:
i see. but since this check cannot be optimized out in the Release build, could you added it in the `else` branch, like
```c++
if (value_replayer) {
  if (value_replayer->get_header_magic() != magic) {
    ceph_abort_msgf("OTree::Extent::Replay: value magic mismatch %x != %x", 
       value_replayer->get_header_magic(), magic);
  }
} else {
  value_replayer = build_value_recorder_by_type(encoded, magic);
  // ....
}
```
what do you think?

Reply:
refined.